Issue #14396: checker-framework.yml should fail execution if report generation fails

relentless-pursuit opened the pull request:
Resolve #14396

This PR enhances the GitHub Actions workflow for the Checker Framework by ensuring execution fails upon detecting errors during report generation.

romani commented on .ci/checker-framework.groovy:
```diff
@@ -112,27 +112,31 @@ private static int checkCheckerFrameworkReport(final String profile) {
         while (lineFromReader != null) {
             println(lineFromReader)
             checkerFrameworkLines.add(lineFromReader)
+            if (lineFromReader.contains("OutOfMemoryError")) {
```
We need any error, can we make code general ?
Can we simply rely on `mvn` exit code?

relentless-pursuit replied:
Okay.

relentless-pursuit replied:
> We need any error, can we make code general ? Can we simply rely on `mvn` exit code?

I am trying to throw an exception the moment it encounters a non-zero exit code, assuming a OOM detected. And, this probably might make the CI/CD to fail. Please let me know your thoughts on it.

romani reviewed (changes requested):
Items

romani reviewed (approved):
I like this a way more

relentless-pursuit commented:
> I like this a way more

how do we test such scenarios in our local?

romani commented:
You can update groovy to start mvn with low memory like `-xmx 512mb`

relentless-pursuit commented:
> You can update groovy to start mvn with low memory like `-xmx 512mb` please google exact syntax.

Sure, thanks.

relentless-pursuit commented:
> You can update groovy to start mvn with low memory like `-xmx 512mb` please google exact syntax.

I tried with reducing the jvm heap size to check if the code i added was running fine. But, even at 128mb, i am not being able to simulate an OOM. The other way i tried was runing a infinite loop, the CI did hint towards a OOM, but, as obvious the lines of code I added to terminal the script wasn't executed. Any alternative or correction needed?

relentless-pursuit commented:
> > You can update groovy to start mvn with low memory like `-xmx 512mb` please google exact syntax.
> 
> @romani, @rnveach I tried with reducing the jvm heap size to check if the code i added was running fine. But, even at 128mb, i am not being able to simulate an OOM. The other way i tried was runing a infinite loop, the CI did hint towards a OOM, but, as obvious the lines of code I added to terminal the script wasn't executed. Any alternative or correction needed?

Even if I am failing to replicate OOM, but, atleast I am being able to register an exit status from the below logs.
```
Caught: java.lang.RuntimeException: Maven process exited with error code: 1
java.lang.RuntimeException: Maven process exited with error code: 1
```

romani commented:
We bumped memory usage for maven to bypass this problem in CI, if we revert such commit we should be able reproduce problem in CI, but instead of weird failure of diff on suppression we should have failed step of report generation

relentless-pursuit commented:
> We bumped memory usage for maven to bypass this problem in CI, if we revert such commit we should be able reproduce problem in CI, but instead of weird failure of diff on suppression we should have failed step of report generation.
> 
> Commit revert might be too much, but revert of lines as [1a041e3#diff-9c5fb3d1b7e3b0f54bc5c4182965c4fe1f9023d449017cece3005d3f90e8e4d8L2938](https://github.com/checkstyle/checkstyle/commit/1a041e316fc8d927fbefa850a194a684b786b2e8#diff-9c5fb3d1b7e3b0f54bc5c4182965c4fe1f9023d449017cece3005d3f90e8e4d8L2938) should bring us back to CI problem to run Checker

@romani After reducing the heap size of the profiles, the checker jobs are failing except `(regex-property-key-compiler-message)` because i didn't change the size for this job specifically. Though, when i look into the failed jobs, i continue to see weird failure of diff on suppression and no failed step of report generation.

relentless-pursuit commented:
@nrmancuso can you provide any insights on the issue? i reduced the jvm heap size and as a consequence checker jobs are failing with suppressions (as mentioned above). however, the intent is to have something more substantial like an exception. Though i have added the code in groovy script, yet not reflecting.

nrmancuso commented on pom.xml:
```diff
@@ -2525,7 +2525,7 @@
               <fork>true</fork>
               <failOnError>false</failOnError>
               <meminitial>1024m</meminitial>
-              <maxmem>8192m</maxmem>
```
Do not reduce maxmem here. We want to allow checker to consume as much memory as possible, while restricting the build system's memory.

romani replied:
this update was for testing only.

relentless-pursuit replied:
> Do not reduce maxmem here. We want to allow checker to consume as much memory as possible, while restricting the build system's memory.

will revert in next push, was for testing. i will instead proceed with an infinite loop and capture the exit code, and see where it goes from there.

nrmancuso replied:
> this update was for testing only.

Yep, I am aware of this.

>We want to allow checker to consume as much memory as possible, while restricting the build system's memory.

nrmancuso replied:
> an infinite loop 

I am not sure what this would get us, other than wasting a massive amount of CI resources and tying up Github actions.

nrmancuso commented on .ci/checker-framework.groovy:
```diff
@@ -127,7 +128,10 @@ private static List<List<String>> getCheckerFrameworkErrors(final String profile
             lineFromReader = reader.readLine()
         }
-        process.waitFor()
+        int exitCode = process.waitFor()
+        if (exitCode != 0) {
+            throw new RuntimeException("Maven process exited with error code: " + exitCode)
+        }
```
Not sure if we can really use this approach, since the build will fail when we have checker compilation errors too, and if we exit early we will not see the diff, right?

nrmancuso replied:
A good approach my to be to figure out if the exit codes for OOMs/other exceptions is different from normal checker compilation failure. A good start to figuring this out might be to log the exit code here and keep track of the results as you go.

romani replied:
yes, I see it now, not good but we can not do anything.
lets all back to catching of error in log.

relentless-pursuit replied:
Got it, thanks.

romani replied:
how do you see out of memmory error at https://github.com/checkstyle/checkstyle/actions/runs/7814021233/job/21314388563?pr=14409, I can not spot it.

nrmancuso replied:
@relentless-pursuit it might be worth researching if we can set a custom exit code upon failure for the maven compiler plugin.

nrmancuso replied:
Update: it looks like the build always returns success?

See: https://github.com/checkstyle/checkstyle/actions/runs/7593228559/job/20683478761, linked builds from the issue itself.

Even when we fail with OOM, build is "successful", so I assume that we return 0 whether we OOM, have normal checker failures, or not. We should explore configuring the maven compiler plugin to fail the build.

relentless-pursuit replied:
> Update: it looks like the build always returns success?
> 
> See: https://github.com/checkstyle/checkstyle/actions/runs/7593228559/job/20683478761, linked builds from the issue itself.
> 
> Even when we fail with OOM, build is "successful", so I assume that we return 0 whether we OOM, have normal checker failures, or not. We should explore configuring the maven compiler plugin to fail the build.
> 
> Edit: ok, this is how we have the plugin set up:
> 
> ```
>               <failOnError>false</failOnError>
> ```
> 
> We could try to do:
> 
> ```
>               <failOnError>true</failOnError>
>               <failOnWarning>false</failOnWarning>
> ```
> 
> And see what happens.

Sure.

nrmancuso reviewed (changes requested):
@relentless-pursuit please keep changes and testing attempts in separate commits to make it clear what changes we made, and how we are testing. I typically make my first commit as the permanent change, then stack testing commits on top until I see the failure I am looking for. We can comment out other CI to save resources while you battle with this; a simple renaming of all other `yml` files for CI configuration is usually enough for this.

romani commented:
@relentless-pursuit , I am not sure on status of this PR, please explicitly request help if you need it.
if you still working, it is ok, we will wait.

relentless-pursuit commented:
> @relentless-pursuit , I am not sure on status of this PR, please explicitly request help if you need it. if you still working, it is ok, we will wait.

Sure, i am just trying to squeeze time and will resume work on this on priority. Thank you for your patience.

romani commented:
> Thank you for your patience.

we can wait for year, thanks for contributing to us.

romani commented:
This is exactly what we need, please remove test commit and lets proceed to ready to merge

nrmancuso commented:
@relentless-pursuit before we drop commits, please do the following so that we can easily revisit and quickly understand what you've done here.

1. Separate your changes here into three commits:

relentless-pursuit commented:
Without any changes to groovy file, we continue to get OOM but with some strange suppressions

https://github.com/checkstyle/checkstyle/actions/runs/7947755703/job/21696975080?pr=14409

After groovy changes, we can handle the OOM properly

https://github.com/checkstyle/checkstyle/actions/runs/7947830206/job/21697133364?pr=14409

relentless-pursuit commented:
> @relentless-pursuit before we drop commits, please do the following so that we can easily revisit and quickly understand what you've done here.
> 
> 1. Separate your changes here into three commits:
>    
>    * one with CI disablement
>    * one with hack to cause failure
>    * one with update to groovy script
> 2. Place a link in the issue to the commit that has the hack to cause CI to fail, with a short description. It is likely that we might need such a hack in the future for something else.
> 3. Please explain how your changes in the groovy script helped us, I do not see the `Maven process exited with error code` message in the output you've shared.
> 
> Thanks a lot for all the work on this.

Done.

relentless-pursuit commented:
> 3\. Please explain how your changes in the groovy script helped us, I do not see the `Maven process exited with error code` message in the output you've shared.

Caught: java.lang.RuntimeException: Maven process exited with error code: 1
java.lang.RuntimeException: Maven process exited with error code: 1
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at checker-framework.getCheckerFrameworkErrors(checker-framework.groovy:132)
	at checker-framework.checkCheckerFrameworkReport(checker-framework.groovy:79)
	at checker-framework.parseArgumentAndExecute(checker-framework.groovy:34)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at checker-framework.run(checker-framework.groovy:17)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)

nrmancuso commented on pom.xml:
```diff
@@ -2933,9 +2933,9 @@
-              <failOnError>false</failOnError>
-              <meminitial>1024m</meminitial>
-              <maxmem>8192m</maxmem>
+              <failOnError>true</failOnError>
```
How will the maven execution fail without this?

relentless-pursuit replied:
If I understood it right, the final commit must have the changes of groovy including `failOnError` be true for all checkers.

nrmancuso replied:
> If I understood it right, the final commit must have the changes of groovy including `failOnError` be true for all checkers.

Yes, that’s right. Right now, we only have one updated, and it is in the “hack” to cause OOM commit.

relentless-pursuit replied:
Thank you for the review. I have made the changes. Please let me know if they are right.

nrmancuso replied:
Ok, let's drop other commits and prepare for merge.

relentless-pursuit replied:
> Ok, let's drop other commits and prepare for merge.

Done.

relentless-pursuit replied:
@nrmancuso, I couldn't find anything substantial in the CI failures. Can you help me point what is leading to the failure?

nrmancuso replied:
@relentless-pursuit codenarc is failing in semaphore:
`Violation: Rule=ThrowRuntimeException P=2 Line=132 Msg=[The type RuntimeException should not be thrown] Src=[throw new RuntimeException("Maven process exited with error code: " + exitCode)]`

nrmancuso replied:
> `System.exit(1)`

After thinking on this, this is a hack that we should probably avoid. I also have doubts about whether we would execute the `finally` block below this to close the reader. @relentless-pursuit let's keep it the way you had it and find a more specific exception class as recommended by codenarc: https://codenarc.org/codenarc-rules-exceptions.html#:~:text=Checks%20for%20throwing%20an%20instance%20of%20java.lang.RuntimeException.%20Throw%20an%20instance%20of%20a%20more%20specific%20exception%20subclass%20instead.

relentless-pursuit replied:
> https://codenarc.org/codenarc-rules-exceptions.html#:~:text=Checks%20for%20throwing%20an%20instance%20of%20java.lang.RuntimeException.%20Throw%20an%20instance%20of%20a%20more%20specific%20exception%20subclass%20instead

Sure, I was cross-checking and it didn't give me OOM error.

nrmancuso replied:
Let's do `IllegalStateException`

relentless-pursuit replied:
> Let's do `IllegalStateException`

Done. I also checked with the hack, and it seems to be working fine. https://github.com/checkstyle/checkstyle/actions/runs/7974597699/job/21770876534?pr=14409

nrmancuso commented:
@relentless-pursuit thanks for working on this, I know these sorts of issues can get pretty involved!

relentless-pursuit commented:
> @relentless-pursuit thanks for working on this, I know these sorts of issues can get pretty involved!

Always happy to help Checkstyle. And, thanks to you and other maintainers for always being patient with the contributors like me.

romani reviewed (approved):
Ok to merge, elegant fix.
